### ISSUE TITLE
Add vscode.languages.registerSnippetProvider API

### ISSUE THREAD

**Issue by dsanders11** [microsoft/vscode#111158]
Currently snippets are statically provided by an extension in `package.json`, and can only limit when they're available to the language ID of the file. This makes it difficult not to pollute the global snippet namespace, and extensions need to be weary of adding too many snippets - snippets relevant only to C++ header files can't be separated out from .cpp files since the language ID is the same.

This limits the usefulness of snippets provided by extensions. Extensions may only want to make snippets available under certain circumstances, like providing lots of snippets relevant to writing tests when the user is in a test file, but not at other times. Extensions may also only be enabled if certain conditions are met, and if the extension is not enabled it likely does not want to be contributing snippets. Extensions may want to provide configuration settings which enable or disable snippets, or use configuration settings to change which snippets are shown. For example, I'd like the built-in Markdown snippet for "unordered list" to use asterisks, not dashes, but it's static and hardcoded to dashes. With a snippet provider this could be easily tied to a configuration setting.

I'm proposing `vscode.languages.registerSnippetProvider`, which is similar to `vscode.languages.registerCompletionItemProvider`. I realize there's `vscode.SnippetString` for use with `vscode.CompletionItem`,  but this proposal has some key differences for snippets:

* Provided snippets would be available in the "Insert Snippet" command
* `provideSnippets` would be given the current selections in the editor, allowing more advanced behavior like those requested in #10561
  * `vscode.CompletionItemProvider` only provides the current position, since completions aren't available when text is selected
  * Snippets support selection (`TM_SELECTED_TEXT`), so this is an inherent shortcoming of providing snippets via `vscode.CompletionItemProvider` - you can't provide snippets which use selections because the snippet won't be accessible
* `resolveSnippet` would allow deferring resolving the `body` for a snippet, where as `vscode.CompletionItemProvider` does not support changing the `insertText` in `resolveCompletionItem`
  * This allows for quickly providing applicable snippet names and descriptions, and allows slower computation of the actual snippet body once the snippet has been selected. While `provideCompletionItems` was suggested as a solution to #10561, it has the shortcoming that any advanced behavior for snippets has to be done at the time they are listed, rather than when they are used.
* No need to muck around with `triggerCharacters` trying to match the default snippet behavior of when snippets are suggested
  * Can continue to use `prefix` like in static snippets and get the standard snippet behavior
* No need to muck around with `vscode.CompletionItem.range`
  * Can continue to use `prefix` like in static snippets and get the standard snippet behavior

Proposed API addition:

```typescript
export namespace languages {
    export function registerSnippetProvider(
        selector: vscode.DocumentSelector,
        provider: vscode.SnippetProvider
    ): vscode.Disposable;
}

export class Snippet {
    name: string;
    description?: string;
    prefix?: string | string[];
    body?: vscode.SnippetString;
}

export interface SnippetProvider<T extends Snippet> {
    provideSnippets(
        document: vscode.TextDocument,
        position: vscode.Position,
        selections: vscode.Selections[],
        token: vscode.CancellationToken,
    ): vscode.ProviderResult<T[]>;

    resolveSnippet?(
        snippet: T,
        token: vscode.CancellationToken
    ): vscode.ProviderResult<T>;
}
```

**Comment by dsanders11** [comment]
@jrieken, can you give a summary of what you don't like, to go with the thumbs down?

**Comment by jrieken** [comment]
I don't see a point in duplicating this since it's a subset of the completion provider API

**Comment by dsanders11** [comment]
Well, all of the bullet points list the reasons why the completion provider API isn't a good fit. Like not showing up in "Insert Snippets" and not being able to use selections since completions aren't shown with selections.

I switched some snippets for my extension to the completion provider API because I didn't want them polluting the global namespace. Re-implementing them in the completion provider is the *reason* I created this issue, because the shortcomings became quite clear. You end up having to re-create a lot of the features of snippets to get the same functionality as the static snippets, especially when it comes to matching the behavior of when they're triggered. You end up just re-implementing `SnippetCompletionProvider` from in the VS Code codebase in your extension.
